File: src/main/scala/de/bst/five_cities/utils/CreateDoc.scala
package de.bst.five_cities.utils

import java.io.PrintWriter

import scala.io.Source

object CreateDoc {
  def main(args: Array[String]) {
    generateDoc("src/test/scala/de/bst/five_cities/FiveCitiesSparkTutorialTest.scala", "README.md")
  }

  def generateDoc(source: String, target: String) {
    val lines = (Source fromFile source).getLines map (BlockOfLines(_))
    val document = ((List[BlockOfLines]() /: lines)
      ((l, b) => if(l.isEmpty || b.blockType != l.last.blockType) l :+ b else l.init :+ (l.last merge b))
      map (_.format) mkString "\n")
    new PrintWriter("README.md") {
      write(document)
      close
    }
  }

  private object BlockOfLines {
    val TYPE_CODE = "CODE"
    val TYPE_TEXT = "TEXT"
    private val COMMENT_PATTERN = "^\\s*// ?(.*)$".r
    private val HIDE_PATTERN = "^\\s*//\\s*!HIDE\\s*$".r

    def apply(line: String): BlockOfLines =
      new BlockOfLines(line :: Nil, typeOf(line), HIDE_PATTERN findFirstIn line isDefined)

    def typeOf(line: String): String =
      if (HIDE_PATTERN findFirstIn line isDefined) TYPE_CODE
      else if (COMMENT_PATTERN findFirstIn line isDefined) TYPE_TEXT
      else TYPE_CODE
  }

  private class BlockOfLines(var lines: List[String], val blockType: String, var hide: Boolean) {
    def merge(other: BlockOfLines) = {
      lines = lines ::: other.lines
      this
    }

    override def toString: String = s"${blockType.toString}\n\t${lines mkString "\n\t"}"

    def format: String = {
      import BlockOfLines._
      blockType match {
        case TYPE_CODE => if (!hide) s"```scala\n${lines mkString "\n"}```" else ""
        case TYPE_TEXT => lines map {
          l =>
            val matcher = COMMENT_PATTERN findFirstMatchIn l
            if (matcher.isDefined) matcher.get group 1 else l
        } mkString "\n"
      }
    }
  }
}

File: src/main/java/de/bst/five_cities/utils/CreateDoc.java
package de.bst.five_cities.utils;

import static org.apache.commons.lang3.StringUtils.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.LinkedList;
import java.util.function.BiConsumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;
import java.util.stream.Stream.Builder;

public class CreateDoc {

	public static void main(String[] args) throws Exception {
		generateDoc("src/test/java/de/bst/five_cities/FiveCitiesSparkTutorialTest.java", "README.md");
	}

	public static void generateDoc(String source, String target) throws IOException {
		Stream<String> documentLines = Files.lines(Paths.get(source)) //
				.map(line -> new BlockOfLines(line)) //
				.collect(//
						() -> new LinkedList<>(), //
						CreateDoc::mergeLinesOfSameTypeToBlocks, //
						dontCombine())
				.stream() //
				.flatMap(BlockOfLines::format);

		Files.write(Paths.get(target), (Iterable<String>) documentLines::iterator);
	}

	private static void mergeLinesOfSameTypeToBlocks(LinkedList<BlockOfLines> blocks, BlockOfLines block) {
		if (!blocks.isEmpty() && (block.type == blocks.getLast().type))
			blocks.getLast().merge(block);
		else
			blocks.add(block);
	}

	private static class BlockOfLines {

		static enum LineType {
			CODE, TEXT;

			static LineType of(String line) {
				if (HIDE_PATTERN.matcher(line).matches())
					return CODE;
				else if (COMMENT_PATTERN.matcher(line).matches())
					return TEXT;
				else
					return CODE;
			}

		}

		private static final Pattern COMMENT_PATTERN = Pattern.compile("^\\s*// ?(.*)$");
		private static final Pattern HIDE_PATTERN = Pattern.compile("^\\s*//\\s*!HIDE\\s*$");

		LinkedList<String> lines = new LinkedList<>();
		LineType type;
		boolean hide = false;

		public BlockOfLines(String line) {
			this.lines.add(line);
			this.type = LineType.of(line);
			this.hide = HIDE_PATTERN.matcher(line).matches();
		}

		public void merge(BlockOfLines other) {
			lines.addAll(other.lines);
		}

		@Override
		public String toString() {
			return type.toString() + "\n\t" + join(lines, "\n\t");
		}

		private Stream<String> format() {
			switch (type) {
			case CODE:
				Builder<String> builder = Stream.<String> builder();
				if (!hide) {
					builder.add("```java");
					lines.forEach(l -> builder.add(l));
					builder.add("```");
				}
				return builder.build();
			case TEXT:
				return lines //
						.stream() //
						.map(line -> {
							Matcher matcher = COMMENT_PATTERN.matcher(line);
							return matcher.matches() ? matcher.group(1) : line;
						});
			default:
				throw new IllegalArgumentException();
			}
		}

	}

	private static <A> BiConsumer<A, A> dontCombine() {
		return (a, b) -> {
			throw new UnsupportedOperationException();
		};
	}

}
